lib/seahorn/Bmc.cc

```diff
@@ -130,205 +129,10 @@ raw_ostream &BmcEngine::toSmtLib(raw_ostream &out) {
-
-    for (auto &I : BB) {
-      if (const DbgValueInst *dvi = dyn_cast<DbgValueInst>(&I)) {
-        if (dvi->getValue() && dvi->getVariable()) {
-          if (out.has_colors())
-            out.changeColor(raw_ostream::RED);
-          DILocalVariable *var = dvi->getVariable();
-          out << "  " << var->getName() << " = ";
-          if (dvi->getValue()->hasName())
-            out << dvi->getValue()->getName();
-          else
-            out << *dvi->getValue();
-          out << "\n";
-          out.resetColor();
-        }
-        continue;
-      }
-
-      bool print_inst = true;
-      bool shadow_mem = false;
-      if (const CallInst *ci = dyn_cast<CallInst>(&I)) {
-        Function *f = ci->getCalledFunction();
-        if (f && f->getName().equals("seahorn.fn.enter")) {
-          if (ci->getDebugLoc()) {
-            if (DISubprogram *fnScope =
-                    getDISubprogram(ci->getDebugLoc().getScope()))
-              out << "enter: " << fnScope->getName() << "\n";
-          }
-          continue;
-        } else if (f && f->getName().equals("shadow.mem.init")) {
-#if 0
-          // disabling since this is not supported by non-legacy
-          // OperationalSemantics
-          if (out.has_color())
-            out.changeColor(raw_ostream::RED);
-          int64_t id = shadow_dsa::getShadowId(ci);
-          assert(id >= 0);
-          Expr memStart = m_bmc.sem().memStart(id);
-          Expr u = eval(loc, memStart);
-          if (u)
-            out << "  " << *memStart << " " << *u << "\n";
-          Expr memEnd = m_bmc.sem().memEnd(id);
-          u = eval(loc, memEnd);
-          if (u)
-            out << "  " << *memEnd << " " << *u << "\n";
-          out.resetColor();
-#endif
-          print_inst = false;
-          shadow_mem = true;
-        } else if (f && f->getName().equals("shadow.mem.store")) {
-          shadow_mem = true;
-        }
-      } else if (isa<PHINode>(I)) {
-        shadow_mem = I.hasMetadata("shadow.mem");
-      }
-
-      if (print_inst) {
-        out << I << "\n";
-      }
-
-      Expr v = eval(loc, I);
-      if (!v)
-        continue;
-
-      if (out.has_colors())
-        out.changeColor(raw_ostream::RED);
-      if (DumpHex && shadow_mem) {
-        using HD = expr::hexDump::HexDump;
-        using SHD = expr::hexDump::StructHexDump;
-        out << "  %" << I.getName() << "\n";
-
-        if (isOp<MK_STRUCT>(v)) {
-          out << SHD(v);
-        } else {
-          out << HD(v);
-        }
-
-      } else
-        out << "  %" << I.getName() << " " << *v;
-      const DebugLoc &dloc = I.getDebugLoc();
-      if (dloc) {
-        if (out.has_colors())
-          out.changeColor(raw_ostream::CYAN);
-        out << " [" << (*dloc).getFilename() << ":" << dloc.getLine() << "]";
-      }
-      out << "\n";
-      out.resetColor();
-    }
-  }
-  out << "End trace\n";
-  return out;
+  return ZBmcTraceTy(*this, std::make_shared<ZModel<EZ3>>(model));
```

Comment:
This is potentially problematic. Double check ownership.

`model` from the `m_smt_solver` is already a smart pointer, wrapping it in another smart pointer is at the very least annoying. Plus the allocation is shifted to the heap...

Reply:
`m_smt_solver.getModel()` returns a `ZModel<EZ3>` object/reference, which is why I wanted to wrap it around `ZModelRef` in the first place; perhaps I should initialize `ZBmcTraceTy::m_model` with the pointer of `m_smt_solver.getModel()` directly?

Reply:
ZModel is a smart pointer. You can copy it by value.